File: src/main/java/org/example/stack/ValidParentheses.java
package org.example.stack;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class ValidParentheses {
    public boolean isValid(String s) {
        Map<Character, Character> map = new HashMap<Character, Character>() {{
            put('}', '{');
            put(']', '[');
            put(')', '(');
        }};

        Stack<Character> stack = new Stack<>();

        for (char c: s.toCharArray()) {
            if (map.containsKey(c)) {
                if (!stack.isEmpty() && stack.peek() == map.get(c)) {
                    stack.pop();
                } else
                    return false;
            } else {
                stack.push(c);
            }
        }

        return stack.isEmpty();
    }
}

File: src/main/java/org/example/stack/ValidParentheses.py
class Solution(object):
    def isValid(self, s):
        dict = {"}": "{", ")": "(", "]": "["}
        stack = []

        for c in s:
            if c in dict:
                if stack and stack[-1] == dict[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)

        return True if not stack else False 


obj = Solution()
# s = "()"
# s = "()[]{}"
# s = "(]"
s = "((([])))"
print(obj.isValid(s))